AK/BigIntBase.h

```diff
@@ -0,0 +1,566 @@
+constexpr int storage_compare(auto const& operand1, auto const& operand2, bool is_for_inequality)
+{
+    bool sign1 = operand1.is_negative(), sign2 = operand2.is_negative();
+    size_t size1 = operand1.size(), size2 = operand2.size();
+
+    if (sign1 != sign2) {
+        if (sign1)
+            return -1;
+        return 1;
+    }
+
+    NativeWord compare_value = extend_sign(sign1);
+    bool differ_in_high_bits = false;
+
+    if (size1 > size2) {
+        for (size_t i = size1; i-- > size2;)
+            if (operand1[i] != compare_value)
+                differ_in_high_bits = true;
+    } else if (size1 < size2) {
+        for (size_t i = size2; i-- > size1;)
+            if (operand2[i] != compare_value)
+                differ_in_high_bits = true;
+    }
+
+    if (differ_in_high_bits)
+        return (size1 > size2) ^ sign1 ? 1 : -1;
+
+    // FIXME: Using min(size1, size2) in the next line triggers -Warray-bounds on GCC with -O2 and
+    //        -fsanitize=address. I have not reported this.
+    //        Reduced testcase: https://godbolt.org/z/TE3MbfhnE
+    for (size_t i = (size1 > size2 ? size2 : size1); i--;) {
+        auto word1 = operand1[i], word2 = operand2[i];
+
+        if (is_for_inequality) {
+            if (word1 != word2)
+                return 1;
+        } else {
+            if (word1 > word2)
+                return 1;
+            if (word1 < word2)
+                return -1;
+        }
+    }
+    return 0;
+}
+
+enum class BitwiseOperation {
+    AND,
+    OR,
+    XOR,
+    INVERT,
+};
+
+#ifdef AK_COMPILER_GCC
+#    define IGNORE_VECTOR_DEPENDENCIES _Pragma("GCC ivdep")
+#else
+#    define IGNORE_VECTOR_DEPENDENCIES
+#endif
+
+// Requirements:
+//  - !operand1.is_signed && !operand2.is_signed && !result.is_signed (the function will also work
+//    for signed storages but will extend them with zeroes regardless of the actual sign).
+template<BitwiseOperation operation>
+constexpr void storage_compute_bitwise(auto const& operand1, auto const& operand2, auto&& result)
+{
+    size_t size1 = operand1.size(), size2 = operand2.size(), size = result.size();
+
+    IGNORE_VECTOR_DEPENDENCIES
+    for (size_t i = 0; i < size; ++i) {
+        auto word1 = i < size1 ? operand1[i] : 0;
+        auto word2 = i < size2 ? operand2[i] : 0;
+
+        if constexpr (operation == BitwiseOperation::AND)
+            result[i] = word1 & word2;
+        else if constexpr (operation == BitwiseOperation::OR)
+            result[i] = word1 | word2;
+        else if constexpr (operation == BitwiseOperation::XOR)
+            result[i] = word1 ^ word2;
+        else if constexpr (operation == BitwiseOperation::INVERT)
+            result[i] = ~word1;
+        else
+            static_assert(((void)operation, false));
+    }
+}
+
+// See `storage_compute_bitwise` for the signedness requirements.
+template<BitwiseOperation operation>
+constexpr void storage_compute_inplace_bitwise(auto const&, auto const& operand2, auto&& result)
```

Comment:
i guess that thereʼs a reason for this “extraneous” signature? 🤔

Reply:
What's wrong with the signature?

Reply:
Oh, I see.
Yes, there is. All of the functions are called like `storage_*(operand1, operand2, result)` (so that I can nicely generate all of the operators via defines) and inplace versions have `result == operand1`, so `operand1` is not required.

Reply:
(not sure if it what you wrote here is worth putting in a comment in code?)

Reply:
Adding a comment never hurts, will do that